### PR TITLE
Fix #57730

### LINKED ISSUE #57979: Regression from stable to nightly: nested impl trait is not allowed
I have some code that looks a bit like this:

```rust
use std::borrow::Borrow;

pub struct Data<TBody>(TBody);

pub fn collect(_: impl IntoIterator<Item = impl Borrow<Data<impl AsRef<[u8]>>>>) {
    unimplemented!()
}
```

On `1.32.0` (and the current beta `1.33.0-beta.4 2019-01-24 635817b9db20ecdcd036`) this compiles successfully.

On the current nightly `1.34.0-nightly 2019-01-28 d8a0dd7ae88023bd09fa` this fails with:

```
error[E0666]: nested `impl Trait` is not allowed
 --> src/main.rs:5:61
  |
5 | pub fn collect(_: impl IntoIterator<Item = impl Borrow<Data<impl AsRef<[u8]>>>>) {
  |                                            -----------------^^^^^^^^^^^^^^^^--
  |                                            |                |
  |                                            |                nested `impl Trait` here
  |                                            outer `impl Trait`
```

A workaround is just to use generics. Did we mean to change the rules around nested impl trait here?

### PR DESCRIPTION
**Pull request by Zoxc**
cc https://github.com/rust-lang/rust/pull/57730

r? @cramertj

### CODE REVIEW

File: src/test/run-pass/impl-trait/nested.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// run-pass
+#![allow(dead_code)]
+use std::borrow::Borrow;
+use core::fmt::Debug;
+
+fn test(_: impl Borrow<impl Debug>) {}
```
Comment (Centril) [inline src/test/run-pass/impl-trait/nested.rs]: This should definitely not be allowed as written here; There's even a comment saying so in https://github.com/rust-lang/rust/pull/57730/files#diff-1b4892578c33275e3b1ff1fe43b7f2f9.

Instead, this should be moved to [`test/ui/impl-trait/where.rs`](https://github.com/rust-lang/rust/blob/e10cecc25e987487ab8b06d34b1d84bda551a0b6/src/test/ui/impl-trait/where-allowed.rs) as an example of what isn't allowed.

Comment (Centril) [reply src/test/run-pass/impl-trait/nested.rs]: Well not exactly `impl A<impl B>`... it would be nice to have such a test.

**Comment by Centril** [comment]
@Zoxc This is now fixing a PR, not an issue...

**Comment by Zoxc** [comment]
It fixes a bug introduced in the PR. Technically it fixes https://github.com/rust-lang/rust/issues/57979 too, but we may want to keep that open until we actually error on the example there.

**Comment by nikomatsakis** [comment]
Per my comment [here](https://github.com/rust-lang/rust/issues/57979#issuecomment-459387604), this could was never actually *meant* to be accepted. I haven't read the details of this PR yet, so I'm not sure precisely what we should do make it start erroring again -- or maybe we want a future compatibility warning.

**Comment by Zoxc** [comment]
This PR just fixed a bug which causes part of AST validation to be skipped. We definitely want to merge this.

**Comment by Centril** [comment]
@Zoxc Ok; Can you add tests ensuring that https://github.com/rust-lang/rust/issues/57979#issue-404516771 (in return position also) doesn't compile?

**Comment by Zoxc** [comment]
@Centril I'm not sure what you mean. That does compile.

**Comment by nikomatsakis** [comment]
@Zoxc ok, I agree that we ought to land this PR, which looks "clearly correct". We'll want to "fix" https://github.com/rust-lang/rust/issues/57979  too but we can handle that separately I suppose.

@bors r+ rollup

**Comment by nikomatsakis** [comment]
@bors r- 

well, let me investigate the "fix" a second

**Comment by nikomatsakis** [comment]
Hmm, according to my local testing, we *still* see the expected error here, even with this patch applied. I'm pushing a test case, let's see what travis thinks.

**Comment by nikomatsakis** [comment]
r=me if travis is happy. As discussed in the compiler meeting, this PR seems to fix an obvious bug, but I don't think that it actually affects the behavior of https://github.com/rust-lang/rust/issues/57979 at all. I added a regression test for the moment that documents the behavior in that case, since apparently it was untested. Locally, that test passes for me.

**Pull request merged**